packages/medusa/src/services/product-variant-inventory.ts

```diff
@@ -638,28 +638,50 @@ class ProductVariantInventoryService extends TransactionBaseService {
 
   async setVariantAvailability(
     variants: ProductVariant[] | PricedVariant[],
-    salesChannelId: string | string[] | undefined
+    salesChannelId: string | string[] | undefined,
+    variantInventoryMap: Map<string, ProductVariantInventoryItem[]> = new Map()
   ): Promise<ProductVariant[] | PricedVariant[]> {
     if (!this.inventoryService_) {
       return variants
     }
 
+    if (!variantInventoryMap.size) {
+      const variantInventories = await this.listByVariant(
+        variants.map((v) => v.id)
+      )
+
+      variantInventories.forEach((inventory) => {
+        const variantId = inventory.variant_id
+        const currentInventories = variantInventoryMap.get(variantId) || []
+        currentInventories.push(inventory)
+        variantInventoryMap.set(variantId, currentInventories)
+      })
+    }
+
     return await Promise.all(
       variants.map(async (variant) => {
         if (!variant.id) {
           return variant
         }
 
-        if (!salesChannelId) {
-          delete variant.inventory_quantity
+        variant.purchasable = variant.allow_backorder
+
+        if (!variant.manage_inventory) {
+          variant.purchasable = true
           return variant
         }
 
-        // first get all inventory items required for a variant
-        const variantInventory = await this.listByVariant(variant.id)
+        const variantInventory = variantInventoryMap.get(variant.id) || []
 
         if (!variantInventory.length) {
           variant.inventory_quantity = 0
```

Comment:
thought: Should we remove inventory quantity, as it would otherwise seem misleading, that the variant is purchasable yet there's no stock? Similar to the case with no sales channel id.